### DISCUSSION TITLE
What is a good way to remove the 'title' field from json schemas - to make them usable in OpenAI chatcompletions calls?

### DISCUSSION THREAD

**Discussion by zby** [pydantic/pydantic#8504] [Question] [answered]
The json schema used in example OpenAI chatcompletion calls don't have the `title` fields: 

```
from openai import OpenAI
client = OpenAI()

tools = [
  {
    "type": "function",
    "function": {
      "name": "get_current_weather",
      "description": "Get the current weather in a given location",
      "parameters": {
        "type": "object",
        "properties": {
          "location": {
            "type": "string",
            "description": "The city and state, e.g. San Francisco, CA",
          },
          "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
        },
        "required": ["location"],
      },
    }
  }
]
messages = [{"role": "user", "content": "What's the weather like in Boston today?"}]
completion = client.chat.completions.create(
  model="gpt-3.5-turbo",
  messages=messages,
  tools=tools,
  tool_choice="auto"
)
```
(from https://platform.openai.com/docs/api-reference/chat/create?lang=python)
My goal is to generate the `parameters` part of that schema from pydantic classes. To be more precise I want to use one parameter functions with the only parameter being a pydantic class. It is especially useful for nested structures.

This is not a new idea - it is used in many popular libraries - like https://github.com/jxnl/instructor, https://github.com/minimaxir/simpleaichat and https://github.com/langroid/langroid - they all remove the `title` fields recursively.
For example:
```
def _recursive_purge_dict_key(d: Dict[str, Any], k: str) -> None:
    """Remove a key from a dictionary recursively"""
    if isinstance(d, dict):
        for key in list(d.keys()):
            if key == k and "type" in d.keys():
                del d[key]
            else:
                _recursive_purge_dict_key(d[key], k)
```
from https://github.com/langroid/langroid/blob/main/langroid/utils/pydantic_utils.py#L30

This feels like a hack to me. Aren't there any better methods?

I googled for possible answers, found some old documentation and not realizing that it is from an old version I asked a question about using `json_schema_extra` for this in: https://github.com/pydantic/pydantic/discussions/8476. I got some answers - but nothing that would lead me to a solid solution to this problem. `json_schema_extra` works partially - but it also seems to be deprecated. `ConfigDict` that in some tests is suggested as a replacement does not seem to have any option to remove the titles.

**Comment by zby** [comment]
I've learned that this is the extended schema format. The OpenAI examples use a minimal schema format. It is not clear if the minimal schema is better - but at least it should require less tokens.
I've asked about this at the OpenAI forum: https://community.openai.com/t/extended-or-minimal-schemas-for-tool-parameters/578636

**Comment by samuelcolvin** [comment, marked as answer]
No way to disable it currently.

**Comment by LiranYoffe** [comment]
A solution I found to remove titles (might not be bulletproof):
```python
from pydantic.json_schema import GenerateJsonSchema

class GenerateJsonSchemaNoTitles(GenerateJsonSchema):
    def field_title_should_be_set(self, _) -> bool:
        return False

    def _update_class_schema(self, json_schema, cls, config) -> None:
        super()._update_class_schema(json_schema, cls, config)
        json_schema.pop("title", None)

schema_without_titles = SomePydanticClass.model_json_schema(schema_generator=GenerateJsonSchemaNoTitles)
```

Regarding the use case: I actually use it for tool declaration for Gemini and Claude.
OpenAI has a builtin utility for that (kinda): openai.pydantic_function_tool. After calling that utility on all tools you can use:
```python
response = await client.responses.parse(
    model=model,
    input=messages,
    tools=tools_declarations
)
```
to get pydantic objects in the tool calls.

Hope I helped someone...

**Reply by dongyuzheng** [reply]
You helped me. Thank you.
